python/ray/autoscaler/_private/vsphere/node_provider.py

```diff
@@ -0,0 +1,1111 @@
+
+                matched_tags = {}
+                all_tags = {}
+                # If the VM has a tag from tag_filter then select it
+                for (
+                    tag_id
+                ) in self.vsphere_automation_sdk_client.tagging.TagAssociation.list_attached_tags(
+                    yn_id
+                ):
+                    vsphere_vm_tag = self.vsphere_automation_sdk_client.tagging.Tag.get(
+                        tag_id=tag_id
+                    ).name
+                    if ":" in vsphere_vm_tag:
+                        tag_key_value = vsphere_vm_tag.split(":")
+                        tag_key = tag_key_value[0]
+                        tag_value = tag_key_value[1]
+                        if tag_key in filters and tag_value == filters[tag_key]:
+                            matched_tags[tag_key] = tag_value
+
+                        all_tags[tag_key] = tag_value
+
+                # Update the tag cache with latest tags
+                self.tag_cache[vm_id] = all_tags
+
+                if len(matched_tags) == len(filters):
+
+                    power_status = (
+                        self.vsphere_automation_sdk_client.vcenter.vm.Power.get(vm_id)
+                    )
+
+                    # Return VMs in powered-on and creating state
+                    if (
+                        power_status.state == HardPower.State.POWERED_OFF
+                        and all_tags[Constants.VSPHERE_NODE_STATUS]
+                        == Constants.VsphereNodeStatus.CREATING.value
+                        or power_status.state == HardPower.State.POWERED_ON
+                    ):
+                        nodes.append(vm_id)
+                        # refresh cached_nodes with latest information e.g external ip
+                        if vm_id in self.cached_nodes:
+                            del self.cached_nodes[vm_id]
+                            self.cached_nodes[vm_id] = vm
+
+            cli_logger.info(f"Nodes are {nodes}")
+            return nodes
+
+    def is_running(self, node_id):
+        node = self._get_cached_node(node_id)
+        return node.power_state in {HardPower.State.POWERED_ON}
+
+    def is_terminated(self, node_id):
+        node = self._get_cached_node(node_id)
+        return node.power_state not in {HardPower.State.POWERED_ON}
+
+    def node_tags(self, node_id):
+        with self.tag_cache_lock:
+            d1 = self.tag_cache[node_id]
+            return dict(d1)
+
+    def external_ip(self, node_id):
+        # Return the external IP of the VM
+
+        vm = self.vsphere_automation_sdk_client.vcenter.vm.guest.Identity.get(node_id)
+
+        cli_logger.info("Fetch IP {} for VM {}".format(vm.ip_address, vm))
+
+        return vm.ip_address
+
+    def internal_ip(self, node_id):
+        # TODO: Currently vSphere VMs do not show an internal IP. Check IP configuration
+        # to get internal IP too. Temporary fix is to just work with external IPs
```

Comment:
Should we log a warning?

Reply:
I think this doesn't worth that,  we removed the misleading comments.